File: Python/Composite/Gemini 1.5 Flash/round2/code.py
from abc import ABC, abstractmethod

class ATMComponent(ABC):
    @abstractmethod
    def execute(self):
        pass

class ATM(ATMComponent):
    def __init__(self):
        self.balance = 1000
        self.pin_attempts = 0
        self.max_attempts = 3

    def execute(self):
        print("Welcome to the ATM")
        while True:
            print("\nChoose an option:")
            print("1. Check balance")
            print("2. Withdraw")
            print("3. Deposit")
            print("4. Change PIN")
            print("5. Exit")

            choice = input("Enter your choice: ")

            if choice == "1":
                self.check_balance()
            elif choice == "2":
                self.withdraw()
            elif choice == "3":
                self.deposit()
            elif choice == "4":
                self.change_pin()
            elif choice == "5":
                print("Thank you for using the ATM!")
                break
            else:
                print("Invalid choice. Please try again.")

    def check_balance(self):
        print(f"Your current balance is: {self.balance}")

    def withdraw(self):
        amount = int(input("Enter the amount to withdraw: "))
        if amount <= self.balance:
            self.balance -= amount
            print(f"You have withdrawn {amount}.")
            print(f"Your remaining balance is: {self.balance}")
        else:
            print("Insufficient funds.")

    def deposit(self):
        amount = int(input("Enter the amount to deposit: "))
        self.balance += amount
        print(f"You have deposited {amount}.")
        print(f"Your current balance is: {self.balance}")

    def change_pin(self):
        old_pin = input("Enter your current PIN: ")
        if self.authenticate(old_pin):
            new_pin = input("Enter your new PIN: ")
            confirm_pin = input("Confirm your new PIN: ")
            if new_pin == confirm_pin:
                self.pin = new_pin
                print("Your PIN has been changed successfully.")
            else:
                print("PINs do not match. Please try again.")
        else:
            print("Invalid PIN. Please try again.")

    def authenticate(self, pin):
        if pin == "1234":
            self.pin_attempts = 0
            return True
        else:
            self.pin_attempts += 1
            if self.pin_attempts >= self.max_attempts:
                print("Card blocked due to excessive incorrect PIN attempts.")
                return False
            else:
                print("Incorrect PIN. Please try again.")
                return False

class PrintSlip(ATMComponent):
    def __init__(self, atm, transaction_details):
        self.atm = atm
        self.transaction_details = transaction_details

    def execute(self):
        print("\n--- Transaction Slip ---")
        print(f"Date: {self.transaction_details['date']}")
        print(f"Time: {self.transaction_details['time']}")
        print(f"Transaction: {self.transaction_details['transaction']}")
        print(f"Amount: {self.transaction_details['amount']}")
        print(f"Balance: {self.atm.balance}")
        print("-----------------------")

class Transaction(ATMComponent):
    def __init__(self, atm, transaction_type):
        self.atm = atm
        self.transaction_type = transaction_type
        self.transaction_details = {}

    def execute(self):
        if self.transaction_type == "withdraw":
            amount = int(input("Enter the amount to withdraw: "))
            self.transaction_details['transaction'] = "Withdrawal"
            self.transaction_details['amount'] = amount
            self.transaction_details['date'] = "2023-12-12"
            self.transaction_details['time'] = "10:10:10"
            if amount <= self.atm.balance:
                self.atm.balance -= amount
                print(f"You have withdrawn {amount}.")
                print(f"Your remaining balance is: {self.atm.balance}")
            else:
                print("Insufficient funds.")
        elif self.transaction_type == "deposit":
            amount = int(input("Enter the amount to deposit: "))
            self.transaction_details['transaction'] = "Deposit"
            self.transaction_details['amount'] = amount
            self.transaction_details['date'] = "2023-12-12"
            self.transaction_details['time'] = "10:10:10"
            self.atm.balance += amount
            print(f"You have deposited {amount}.")
            print(f"Your current balance is: {self.atm.balance}")
        elif self.transaction_type == "check_balance":
            self.transaction_details['transaction'] = "Balance Inquiry"
            self.transaction_details['amount'] = 0
            self.transaction_details['date'] = "2023-12-12"
            self.transaction_details['time'] = "10:10:10"
            print(f"Your current balance is: {self.atm.balance}")

def main():
    atm = ATM()
    atm.execute()

if __name__ == "__main__":
    main()

File: Java/Composite/Gemini 1.5 Flash/round2/code.java
import java.util.HashMap;
import java.util.Map;

// Abstract Component - Bank Account Component
abstract class BankAccountComponent {
    protected String accountNumber;

    public BankAccountComponent(String accountNumber) {
        this.accountNumber = accountNumber;
    }

    public abstract void displayBalance();
    public abstract void withdraw(double amount);
    public abstract void deposit(double amount);
    public abstract void changePIN(String oldPIN, String newPIN);
    public abstract String getAccountNumber();
}

// Leaf - Checking Account
class CheckingAccount extends BankAccountComponent {
    private double balance;
    private String pin;

    public CheckingAccount(String accountNumber, String pin, double initialBalance) {
        super(accountNumber);
        this.balance = initialBalance;
        this.pin = pin;
    }

    @Override
    public void displayBalance() {
        System.out.println("Account Balance: " + balance);
    }

    @Override
    public void withdraw(double amount) {
        if (balance >= amount) {
            balance -= amount;
            System.out.println("Withdrawal successful. New balance: " + balance);
        } else {
            System.out.println("Insufficient funds.");
        }
    }

    @Override
    public void deposit(double amount) {
        balance += amount;
        System.out.println("Deposit successful. New balance: " + balance);
    }

    @Override
    public void changePIN(String oldPIN, String newPIN) {
        if (oldPIN.equals(this.pin)) {
            this.pin = newPIN;
            System.out.println("PIN changed successfully.");
        } else {
            System.out.println("Incorrect old PIN.");
        }
    }

    public double getBalance() {
        return balance;
    }

    public String getPIN() {
        return pin;
    }
}

// Composite - ATM
class ATM {
    private Map<String, BankAccountComponent> accounts;
    private int pinAttempts;

    public ATM() {
        accounts = new HashMap<>();
        pinAttempts = 0;
    }

    public void addAccount(String accountNumber, String pin, double initialBalance) {
        accounts.put(accountNumber, new CheckingAccount(accountNumber, pin, initialBalance));
    }

    public void authenticate(String accountNumber, String pin) {
        BankAccountComponent account = accounts.get(accountNumber);
        if (account != null && account.getPIN().equals(pin)) {
            System.out.println("Authentication successful.");
            pinAttempts = 0; // Reset attempts
            processTransactions(account);
        } else {
            pinAttempts++;
            if (pinAttempts >= 3) {
                System.out.println("Card retained. Too many incorrect PIN attempts.");
            } else {
                System.out.println("Incorrect PIN. Attempts remaining: " + (3 - pinAttempts));
            }
        }
    }

    private void processTransactions(BankAccountComponent account) {
        while (true) {
            System.out.println("\nSelect an option:");
            System.out.println("1. Display Balance");
            System.out.println("2. Withdraw");
            System.out.println("3. Deposit");
            System.out.println("4. Change PIN");
            System.out.println("5. Exit");

            int choice = 0;
            try {
                choice = Integer.parseInt(System.console().readLine());
            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Please enter a number.");
                continue;
            }

            switch (choice) {
                case 1:
                    account.displayBalance();
                    break;
                case 2:
                    System.out.print("Enter withdrawal amount: ");
                    try {
                        double amount = Double.parseDouble(System.console().readLine());
                        account.withdraw(amount);
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid amount. Please enter a number.");
                    }
                    break;
                case 3:
                    System.out.print("Enter deposit amount: ");
                    try {
                        double amount = Double.parseDouble(System.console().readLine());
                        account.deposit(amount);
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid amount. Please enter a number.");
                    }
                    break;
                case 4:
                    System.out.print("Enter old PIN: ");
                    String oldPIN = System.console().readLine();
                    System.out.print("Enter new PIN: ");
                    String newPIN = System.console().readLine();
                    System.out.print("Confirm new PIN: ");
                    String confirmPIN = System.console().readLine();
                    if (newPIN.equals(confirmPIN)) {
                        account.changePIN(oldPIN, newPIN);
                    } else {
                        System.out.println("New PINs do not match.");
                    }
                    break;
                case 5:
                    System.out.println("Exiting ATM.");
                    return;
                default:
                    System.out.println("Invalid option. Please try again.");
            }
        }
    }
}